### PR TITLE
parametric(ruby): enables span links tests for ruby tracer

### PR DESCRIPTION
**Pull request by mabdinur**
## Motivation

Enabling span link system tests for the ruby tracer

## Changes

- Enables Span Link tests for the ruby tracer
- Adds support for generating span links to the ruby parametric app
  - Implements the `span_add_link` ruby endpoint
  - Implements span_link initialization logic in `start_span` endpoint
  - Adds a mechanism to add span_links on non-active spans
- Clean up
  - Resolves bugs in the span link grpc interface
  - Ensure span link protobufs are consistent with the apm_client proto file

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
    * [ ] To R&P team: locally build and push the image to hub.docker.com 
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
    * [ ] Once merged, add (or remove) it in system-test-dasboard nightly

### CODE REVIEW

**Comment by mabdinur** [comment]
Blocked by: https://github.com/DataDog/dd-trace-rb/pull/3596

File: tests/parametric/conftest.py [hunk 1]
```diff
@@ -654,6 +654,8 @@ def wait_for_num_traces(self, num: int, clear: bool = False, wait_loops: int = 3
                 if num_received == num:
                     if clear:
                         self.clear()
+                    for trace in traces:
```
Comment (mabdinur) [inline tests/parametric/conftest.py:657]: This change was moved here: https://github.com/DataDog/system-tests/pull/2367

File: tests/parametric/test_otel_span_methods.py [hunk 2]
```diff
@@ -449,7 +449,6 @@ def test_otel_span_started_with_link_from_another_span(self, test_agent, test_li
     @missing_feature(context.library < "java@1.26.0", reason="Implemented in 1.26.0")
     @missing_feature(context.library == "golang", reason="Not implemented")
     @missing_feature(context.library == "nodejs", reason="Not implemented")
```
Comment (mabdinur) [inline tests/parametric/test_otel_span_methods.py:451]:
```suggestion
    @missing_feature(context.library == "nodejs", reason="Not implemented")
    @missing_feature(context.library < "ruby@2.0.0", reason="Not implemented")
```

File: tests/parametric/test_otel_span_methods.py [hunk 3]
```diff
@@ -500,7 +499,6 @@ def test_otel_span_started_with_link_from_datadog_headers(self, test_agent, test
     @missing_feature(context.library < "java@1.28.0", reason="Implemented in 1.28.0")
     @missing_feature(context.library == "golang", reason="Not implemented")
     @missing_feature(context.library == "nodejs", reason="Not implemented")
```
Comment (mabdinur) [inline tests/parametric/test_otel_span_methods.py:501]:
```suggestion
    @missing_feature(context.library == "nodejs", reason="Not implemented")
    @missing_feature(context.library < "ruby@2.0.0", reason="Not implemented")
```

File: tests/parametric/test_otel_span_methods.py [hunk 4]
```diff
@@ -548,7 +546,6 @@ def test_otel_span_started_with_link_from_w3c_headers(self, test_agent, test_lib
     @missing_feature(context.library < "java@1.26.0", reason="Implemented in 1.26.0")
     @missing_feature(context.library == "golang", reason="Not implemented")
     @missing_feature(context.library == "nodejs", reason="Not implemented")
```
Comment (mabdinur) [inline tests/parametric/test_otel_span_methods.py:548]:
```suggestion
    @missing_feature(context.library == "nodejs", reason="Not implemented")
    @missing_feature(context.library < "ruby@2.0.0", reason="Not implemented")
```

File: utils/parametric/_library_client.py [hunk 5]
```diff
@@ -222,12 +222,7 @@ def span_add_link(
     ):
         self._session.post(
             self._url("/trace/span/add_link"),
-            json={
-                "span_id": span_id,
-                "parent_id": parent_id,
-                "attributes": attributes or {},
-                "http_headers": http_headers or [],
-            },
+            json={"span_id": span_id, "parent_id": parent_id, "attributes": attributes or {},},
```
Comment (mabdinur) [inline utils/parametric/_library_client.py:225]:
```suggestion
            json={
                "span_id": span_id,
                "parent_id": parent_id,
                "attributes": attributes or {},
                "http_headers": http_headers or [],
            },
```

File: utils/build/docker/ruby/parametric/server.rb [hunk 6]
```diff
@@ -149,11 +158,7 @@ def inject_headers(inject_headers_args, _call)
     find_span(inject_headers_args.span_id)
 
     env = {}
-    if Datadog::Tracing::Contrib::GRPC.respond_to?(:inject)
-      Datadog::Tracing::Contrib::GRPC.inject(Datadog::Tracing.active_trace.to_digest, env)
-    else
-      Datadog::Tracing::Contrib::GRPC::Distributed::Propagation.new.inject!(Datadog::Tracing.active_trace.to_digest, env)
-    end
+    extract_grpc_headers(headers)
```
Comment (mabdinur) [inline utils/build/docker/ruby/parametric/server.rb:161]:
```suggestion
    if Datadog::Tracing::Contrib::GRPC.respond_to?(:inject)
      Datadog::Tracing::Contrib::GRPC.inject(Datadog::Tracing.active_trace.to_digest, env)
    else
      Datadog::Tracing::Contrib::GRPC::Distributed::Propagation.new.inject!(Datadog::Tracing.active_trace.to_digest, env)
    end
```

**Review by robertomonteromiguel — CHANGES_REQUESTED** [review]
Your are not triggering the Test_Span_Links test for current dev version of ruby tracer. Is that what you wanted?

**Comment by mabdinur** [comment]
> Your are not triggering the Test_Span_Links test for current dev version of ruby tracer. Is that what you wanted?

`Test_Span_Links` is set to a minimum version of `2.0.0` and datadog v2.0.0 has not been released: https://github.com/DataDog/dd-trace-rb/releases this could explain why are not running these tests for the current dev version (dev version is likely v2.0.0some_release_candidate). 

I verified locally that these tests pass but should we hold off on merging this change until 2.0.0 is released?

**Review by robertomonteromiguel — APPROVED** [review]

**Comment by robertomonteromiguel** [comment]
> > Your are not triggering the Test_Span_Links test for current dev version of ruby tracer. Is that what you wanted?
> 
> `Test_Span_Links` is set to a minimum version of `2.0.0` and datadog v2.0.0 has not been released: https://github.com/DataDog/dd-trace-rb/releases this could explain why are not running these tests for the current dev version (dev version is likely v2.0.0some_release_candidate) OR maybe we introduced a regression in this PR: #2222.
> 
> I verified locally that these tests pass but should we hold off on merging this change until 2.0.0 is released?

It was only for double check....you can merge it

**Pull request merged**